### PR TITLE
Adding advection schemes into Shallow Water

### PR DESCRIPTION
**Pull request by francispoulin**
This allows the user to use all the different advection schemes in ShallowWaterModel.

The method uses second-order center differencing for pressure.  I tried using fourth-order and didn't get it to work.  If someone has an altenrative means to compute the formula on a staggered grid, I would be happy to compare it with what I found.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Hmmm I'm not really sure how it's derived but is the correct stencil the same one as the `CenteredFourthOrder` advection scheme?

https://github.com/CliMA/Oceananigans.jl/blob/master/src/Advection/centered_fourth_order.jl

Same as: https://mitgcm.readthedocs.io/en/latest/algorithm/adv-schemes.html#centered-fourth-order-advection

**Comment by ali-ramadhan** [comment]
Ah I think tests are failing because of #1262 so you might have to `git rebase master` or `git merge master`.

**Comment by francispoulin** [comment]
> Ah I think tests are failing because of #1262 so you might have to `git rebase master` or `git merge master`.

Thanks @ali-ramadhan .  Conversely, I just updated test_shallow_water_model.jl, the same way you and pushed that. I suspect the tests will pass, as they did for me but we will see shortly.

File: Project.toml
Comment (ali-ramadhan) [inline Project.toml:23]: Ah this seems like leftovers from when we accidently installed Revise.jl in the Oceananigans environment haha.

We probably want to remove it before merging with `]rm Revise`.

Comment (francispoulin) [reply Project.toml]: Sorry about that.  I have since removed it and then pushed the changes.  All the tests pass on my comptuer and hope they will pass on github as well.

**Comment by francispoulin** [comment]
> Hmmm I'm not really sure how it's derived but is the correct stencil the same one as the `CenteredFourthOrder` advection scheme?
> 
> https://github.com/CliMA/Oceananigans.jl/blob/master/src/Advection/centered_fourth_order.jl
> 
> Same as: https://mitgcm.readthedocs.io/en/latest/algorithm/adv-schemes.html#centered-fourth-order-advection

Its hard for me to tell whether these are the correct equations or not.  I will have go to into the details of it to unravel the notation.  But thanks for pointing this out.

**Comment by glwagner** [comment]
Looks slightly different from 

image

from 

https://crd.lbl.gov/assets/pubs_presos/AMCS/ANAG/O4AdvDiff.pdf

The distinction is that you're trying to approximate the _derivative_ of a cell-averaged quantity at the face-averaged location, rather merely interpolating a cell-averaged quantity to face-averaged locations. I think.

**Comment by francispoulin** [comment]
> Looks slightly different from 2.5 in
> 
> image
> 
> from
> 
> https://crd.lbl.gov/assets/pubs_presos/AMCS/ANAG/O4AdvDiff.pdf
> 
> The distinction is that you're trying to approximate the _derivative_ of a cell-averaged quantity at the face-averaged location, rather than merely interpolating a cell-averaged quantity to face-averaged locations. I think.

Thanks @glwagner that is a great resource.  Unfortuantely, my formula was only appropriate for Finite Differences and not Finite Volume.  I will read through this work right now to understand where the formula comes from.  But it seems like the two formulas you quoted would be the right ones for computing the derivatives and averages in a FV context, to yield 4th order.  

This PR doesn't necessarily need this but if it's only a few lines of code, then why not include it.  I hope to be able to make the appropriate changes this morning and do a bit of testing, to convince myself that we do have 4th order accuracy on the pressure terms.

**Comment by francispoulin** [comment]
Update: I used the formulas in the paper to approximate the gradient of pressure.  In a test case that is pressure dominated, I found that the accuracy imporoved by more than what the theory predicted (> 16), but that is very encouraging that we are doing something right.

It then occurred to me that we are  using centre differencing for the divergence of the transport in the evolution equation for the height.  That suggests that we would probably want to use 4-th order approximations to those terms as well.

Question:  Since we are chaging from approximations at faces to approximations at centers, should the formuals change? I don't think they should but wanted to ask,  just in case.

**Comment by ali-ramadhan** [comment]
Referring to https://github.com/CliMA/Oceananigans.jl/blob/master/src/Operators/README I think the formula doesn't change but the indices are just shifted down by one?

Relevant section:
```
The geometry of the staggerd grid used by Oceananigans (sometimes called the C-grid)
is (in one dimension) shown below

face   cell   face   cell   face

        i-1            i
         ↓             ↓
  |      ×      |      ×      |
  ↑             ↑             ↑
 i-1            i            i+1

Difference operators are denoted by a `δ` (`\delta`). Calculating the difference
of a cell-centered quantity c at cell i will return the difference at face i

    δcᵢ = cᵢ - cᵢ₋₁

and so this operation, if applied along the x-dimension, is denoted by `δxᶠᵃᵃ`.

The difference of a face-centered quantity u at face i will return the difference
at cell i

    δuᵢ = uᵢ₊₁ - uᵢ
```

**Comment by francispoulin** [comment]
Thanks @ali-ramadhan.  I agree and confirmed that where the derivative is computed on faces and centres, the following line and the line below.  This should extend to higher orders of course.

https://github.com/CliMA/Oceananigans.jl/blob/ebd7858000c78f7b1754112a68c587e6195e620b/src/Operators/derivative_operators.jl#L18

**Comment by glwagner** [comment]
> Update: I used the formulas in the paper to approximate the gradient of pressure. In a test case that is pressure dominated, I found that the accuracy imporoved by more than what the theory predicted (> 16), but that is very encouraging that we are doing something right.

Note that "4th-order" refers to the rate of convergence of the scheme as the grid spacing is reduced. For a particular resolution (and assuming that you are in the "asymptotic regime of convergence", the improvement gained from higher-order scheme involves a constant that's specific to the problem (thus you won't have a theoretical prediction for improvement at fixed resolution, I don't think). Empirically we usually find that higher-order schemes reduce the constant by quite a bit! Which is good --- since we often run marginally resolved problems, improving the constant (rather than the slope/rate of convergence) turns out to be the most important benefit of a high-order scheme.

The plot is super busy, but the point is hopefully illustrated: higher-order advection schemes not only converge _more quickly_ to the exact solution as resolution is increased but _also_ have reduced error at a _fixed_ resolution:

**Comment by glwagner** [comment]
> Update: I used the formulas in the paper to approximate the gradient of pressure. In a test case that is pressure dominated, I found that the accuracy imporoved by more than what the theory predicted (> 16), but that is very encouraging that we are doing something right.

Note that "4th-order" refers to the rate of convergence of the scheme as the grid spacing is reduced. For a particular resolution (and assuming that you are in the "asymptotic regime of convergence", the improvement gained from higher-order scheme involves a constant that's specific to the problem (thus you won't have a theoretical prediction for improvement at fixed resolution, I don't think). Empirically we usually find that higher-order schemes reduce the constant by quite a bit! Which is good --- since we often run marginally resolved problems, improving the constant (rather than the slope/rate of convergence) turns out to be the most important benefit of a high-order scheme.

The plot is super busy, but the point is hopefully illustrated: higher-order advection schemes not only converge _more quickly_ to the exact solution as resolution is increased but _also_ have reduced error at a _fixed_ resolution:

image

**Comment by glwagner** [comment]
Re conversions from cell-averaged quantities to face-averaged quantities and vice versa --- I believe @ali-ramadhan is correct in that you just need to take care to use the correct indexing convention. The formulas should otherwise be identical, because there is nothing special about a face-average or cell-average other than the half-cell shifted location / indexing convention.

File: src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:5]:
```suggestion
@inline squared(i, j, k, grid, ϕ) = @inbounds ϕ[i, j, k]^2
```

Comment (francispoulin) [reply src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:5]: Happy to make the change, but I admit I don't know why we use @inbounds sometimes and not others.

**Comment by glwagner** [comment]
@francispoulin I noticed you are adding convergence tests. We also have convergence tests under `validation` here:

https://github.com/CliMA/Oceananigans.jl/tree/master/validation/convergence_tests

and it could make sense to contribute your convergence tests there.

The tests are used to generate figures for documentation:

https://clima.github.io/OceananigansDocumentation/stable/validation/convergence_tests/

@ali-ramadhan can comment on how convergence tests might be added to our CI / automated testing suite.

Is this PR just intended just to add advection schemes, or it is attempting to add higher-order operators for other terms as well? It'd be great to have higher-order finite volume operators for diffusion terms, for example. Does it make sense to contribute these in a separate PR and consider how they might be used in diffusion operators (for example)?

**Comment by francispoulin** [comment]
> @francispoulin I noticed you are adding convergence tests. We also have convergence tests under `validation` here:
> 
> https://github.com/CliMA/Oceananigans.jl/tree/master/validation/convergence_tests
> 
> and it could make sense to contribute your convergence tests there.
> 
> The tests are used to generate figures for documentation:
> 
> https://clima.github.io/OceananigansDocumentation/stable/validation/convergence_tests/
> 
> @ali-ramadhan can comment on how convergence tests might be added to our CI / automated testing suite.
> 
> Is this PR just intended just to add advection schemes, or it is attempting to add higher-order operators for other terms as well? It'd be great to have higher-order finite volume operators for diffusion terms, for example. Does it make sense to contribute these in a separate PR and consider how they might be used in diffusion operators (for example)?

Thanks @glwagner for the post.  I think it makes sense for this to just include updating the advection, since that's already a bit step.  I am sorry for no doing this sooner.  I got distracted thinking it would just be a little more but it turns out to be more than I thought.  I will clean up the branch now and then maybe you can approve the update?

Next week I will work on getting higher order methods working for ShallowWaterModels.

**Comment by glwagner** [comment]
> Next week I will work on getting higher order methods working for ShallowWaterModels.

That would be awesome!

Does it also make sense to work on producing some validation tests for `ShallowWaterModel`? While we can test that operators are implemented correctly in isolation, we won't understand whether everything is integrated into `ShallowWaterModel` until we have some dynamics tests and larger validation runs that reproduce results in the literature, right? Or do we need these higher-order operators to use `ShallowWaterModel` at all?

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:2]: Extra whitespace here

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:16]: Do we need to use high-order interpolation for `h` here? It might be inconsistent (or just effectively second-order) if we use high-order intepolation for `uh` and `vh` but then just second-order interpolation for `h`.

Two-way high-order intepolation for `h` is slightly annoying, but doable. You'll need to use functions like `_symmetric_interpolate_yᵃᶠᵃ`, etc:

https://github.com/CliMA/Oceananigans.jl/blob/master/src/Advection/centered_advective_fluxes.jl#L15

As a side note, its a bit messy (but makes sense) that we have special interpolation functions dispatching on the _advection_ scheme. High-order interpolation isn't really specific to advection; it's just that we only use it for advection currently. A cleaner design would separate order of intepolation from the concept of advection, and perhaps provide generic interpolation and differentiation operators for n-th order (like what we have for second-order interpolation and differentiation). Food for thought.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:19]: Similar problem as above regarding high-order interpolation of `uh`, `vh`, combined with second-order interpolation of `h`.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:37]: Bit of extra whitespace down here.

File: src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:8]: This function could probably be defined on a single line.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [inline src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:27]: Extra whitespace here

**Review by glwagner — APPROVED** [review]
This looks great. It could make sense to add a test that loops over a few advection schemes to ensure that the advection scheme in `ShallowWaterModel` can be user-controlled without an error getting thrown. It's ok to hold off for a future PR though since we're still in "experimental" stage.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:16]: Could make sense to save fixing `h` interpolation for a future PR.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:37]: Echoing an above comment: could make sense to save fixing `h` interpolation for a future PR.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (glwagner) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:19]: Echoing an above comment: could make sense to save fixing `h` interpolation for a future PR.

**Comment by francispoulin** [comment]
> > Next week I will work on getting higher order methods working for ShallowWaterModels.
> 
> That would be awesome!
> 
> Does it also make sense to work on producing some validation tests for `ShallowWaterModel` for a next PR? While we can test that operators are implemented correctly in isolation, we won't understand whether everything is integrated into `ShallowWaterModel` until we have some dynamics tests and larger validation runs that reproduce results in the literature, right? Or do we need these higher-order operators to use `ShallowWaterModel` at all?
> 
> To validate the generic implementation of higher-order operators, I think we can implement high-order diffusion and reuse some of the convergence validation tests for `IncompressibleModel`, like the forced flow fixed slip test. When `ShallowWaterModel` has similar validation tests we can also do the same for `ShallowWaterModel`.

I think that getting some tests is a good next step.  I'll see what kind of tests we have now for IncompressibleModel and will put together a few tests for ShallowWaterModel

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:2]: Removed space

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:16]: At the moment we are just using 2nd order approximation to the gradient of pressure, so using a 2nd order interolation for h is fine.  However, if we want higher order approximations for pressure, then this is something that needs to be sorted out.  This is not conceptually difficult but it will need to be tested and confirmed that it's worth the extra effort.  But that will be for another PR.

Short answer, yes.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:19]: Agreed

File: src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:8]: Done

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl:27]: Good catch.  Fixed.

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl]: removed

File: src/Models/ShallowWaterModels/shallow_water_advection_operators.jl
Comment (francispoulin) [reply src/Models/ShallowWaterModels/shallow_water_advection_operators.jl]: Removed

**Pull request merged**
